crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs

```diff
@@ -0,0 +1,554 @@
+            commit_resolver,
+            permit,
+        };
+
+        if let Err(err) = self.tx.send(inflight_append) {
+            // if we failed to push this to be processed by the connection reactor task
+            // then we need to notify the caller
+            err.0
+                .commit_resolver
+                .error(AppendError::retryable(NetworkError::ConnectionClosed));
+        }
+    }
+
+    /// Handle all [`Appended`] responses
+    ///
+    /// This task will run until the [`AppendStream`] is dropped. Once dropped
+    /// all pending commits will be resolved with an error. it's up to the enqueuer
+    /// to retry if needed.
+    async fn handle_appended_responses(
+        known_global_tail: TailOffsetWatch,
+        connection: WeakConnection,
+        mut rx: mpsc::UnboundedReceiver<RemoteInflightAppend>,
+    ) -> anyhow::Result<()> {
+        let mut closed = std::pin::pin!(connection.closed());
+
+        // handle all rpc tokens in a loop, this loop will
+        // break only with a break reason.
+        // this reason can be one of the following:
+        // - Shutdown error
+        // - Terminal AppendError
+        // - Connection closed
+        let reason = loop {
+            let inflight = match rx.recv().await {
+                Some(inflight) => inflight,
+                None => {
+                    break BreakReason::Shutdown(ShutdownError);
+                }
+            };
+
+            tokio::select! {
+                appended = inflight.rpc_token.recv() => {
+                    let appended = match appended {
+                        Ok(appended) => appended,
+                        Err(err) => {
+                            break BreakReason::Shutdown(err);
+                        }
+                    };
+
+                    if let Err(err) = Self::resolve(&known_global_tail, inflight.commit_resolver, appended.into_body()) {
+                        break BreakReason::AppendError(err);
+                    }
+                },
+                _ = &mut closed => {
+                    break BreakReason::ConnectionClosed;
+                }
+            }
+        };
+
+        // close channel to stop any further appends calls on the same connection
+        rx.close();
+
+        // drain and resolve ALL pending appends on this connection.
+        while let Some(inflight) = rx.recv().await {
+            match reason {
+                BreakReason::Shutdown(_) => {
+                    inflight
+                        .commit_resolver
+                        .error(AppendError::terminal(ShutdownError));
+                }
+                BreakReason::ConnectionClosed => inflight
+                    .commit_resolver
+                    .error(AppendError::retryable(NetworkError::ConnectionClosed)),
+                BreakReason::AppendError(ref err) => {
+                    inflight.commit_resolver.error(err.clone());
+                }
+            }
+        }
```

Comment:
Is it correct that the `RemoteInflightAppends` on `rx` don't necessarily have the same order as the `Appends` that are being sent to the remote sequencer (e.g. if there are two independent writers)? If this is true, then `rx` could contain some `RIAppends` which received a positive response from the sequencer. Maybe this is a micro optimization and it is better to state here why it is ok to fail these requests instead.